includes/express-checkout/class-wc-payments-express-checkout-button-helper.php

```diff
@@ -232,4 +241,153 @@ private function get_quantity() {
+	 * @return boolean
+	 */
+	public function is_available_at( $location, $option_name ) {
+		$available_locations = $this->gateway->get_option( $option_name );
+		if ( $available_locations && is_array( $available_locations ) ) {
+			return in_array( $location, $available_locations, true );
+		}
+
+		return false;
+	}
+
+	/**
+	 * The settings for the `button` attribute - they depend on the "grouped settings" flag value.
+	 *
+	 * @return array
+	 */
+	public function get_button_settings() {
+		$button_type = $this->gateway->get_option( 'payment_request_button_type' );
+		return [
+			'type'         => $button_type,
+			'theme'        => $this->gateway->get_option( 'payment_request_button_theme' ),
+			'height'       => $this->get_button_height(),
+			// Default format is en_US.
+			'locale'       => apply_filters( 'wcpay_payment_request_button_locale', substr( get_locale(), 0, 2 ) ),
+			'branded_type' => 'default' === $button_type ? 'short' : 'long',
+			'size'         => $this->gateway->get_option( 'payment_request_button_size' ),
+			'context'      => $this->get_button_context(),
+		];
+	}
+
+	/**
+	 * Gets the context for where the button is being displayed.
+	 *
+	 * @return string
+	 */
+	public function get_button_context() {
+		if ( $this->is_product() ) {
+			return 'product';
+		}
+
+		if ( $this->is_cart() ) {
+			return 'cart';
+		}
+
+		if ( $this->is_pay_for_order_page() ) {
+			return 'pay_for_order';
+		}
+
+		if ( $this->is_checkout() ) {
+			return 'checkout';
+		}
+
+		return '';
+	}
+
+	/**
+	 * Gets the button height.
+	 *
+	 * @return string
+	 */
+	public function get_button_height() {
+		$height = $this->gateway->get_option( 'payment_request_button_size' );
+		if ( 'medium' === $height ) {
+			return '48';
+		}
+
+		if ( 'large' === $height ) {
+			return '56';
+		}
+
+		// for the "default"/"small" and "catch-all" scenarios.
+		return '40';
+	}
+
+	/**
+	 * Get product from product page or product_page shortcode.
+	 *
+	 * @return WC_Product|false|null Product object.
+	 */
+	public function get_product() {
+		global $post;
+
+		if ( is_product() ) {
+			return wc_get_product( $post->ID );
+		} elseif ( wc_post_content_has_shortcode( 'product_page' ) ) {
+			// Get id from product_page shortcode.
+			preg_match( '/\[product_page id="(?<id>\d+)"\]/', $post->post_content, $shortcode_match );
+			if ( isset( $shortcode_match['id'] ) ) {
+				return wc_get_product( $shortcode_match['id'] );
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Returns true if the provided WC_Product is a subscription, false otherwise.
+	 *
+	 * @param WC_Product $product The product to check.
+	 *
+	 * @return bool  True if product is subscription, false otherwise.
+	 */
+	public function is_product_subscription( WC_Product $product ): bool {
+		return 'subscription' === $product->get_type()
+			|| 'subscription_variation' === $product->get_type()
+			|| 'variable-subscription' === $product->get_type();
+	}
```

Comment:
How this method is different from calling `WC_Subscriptions_Product::is_subscription( $product );`? It seems to me they achieve the same so maybe it's worth removing this method all and instead call `WC_Subscriptions_Product::is_subscription( $product );` in `class-wc-payments-woopay-button-handler.php::283`? Please let me know if I'm missing something :)

Reply:
My understanding is that `is_product_subscription()` was added to cover both Woo Subscription products as well as WooPayments subscription products, whereas `WC_Subscriptions_Product::is_subscription( $product )` requires Woo Subscriptions and will only account for Woo Subscriptions products. I've inferred this from the [PR that added the function](https://github.com/Automattic/woocommerce-payments/pull/6133) with the key piece being: 

>Note
"Subscription product" here refers to both WC Subscriptions products and WCPay Subscriptions products.

Reply:
Thanks for the explanation, all makes sense!